backend/src/database/migrations/20221012051800-add-api-key-hash-column-to-user.js

```diff
@@ -0,0 +1,58 @@
+'use strict'
+
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    await queryInterface.addColumn('users', 'api_key_hash', {
+      type: Sequelize.DataTypes.STRING(255),
+      allowNull: true,
+    })
+    await queryInterface.createFunction(
+      'sync_api_key_with_api_key_hash',
+      [],
+      'trigger',
+      'plpgsql',
+      `
+      NEW.api_key_hash = NEW.api_key;
+      RETURN NEW;
+      `,
+      [],
+      { force: true }
+    )
+    // this trigger is needed to ensure that any users that generate new api keys AFTER the backfill will be synced to api_key_hash
+    await queryInterface.sequelize.query(
+      'CREATE TRIGGER sync_api_key_with_api_key_hash_trigger BEFORE INSERT OR UPDATE ON users FOR EACH ROW EXECUTE PROCEDURE sync_api_key_with_api_key_hash();'
+    )
+    await queryInterface.sequelize.query(
+      'UPDATE users SET api_key_hash = api_key;'
+    )
+    await queryInterface.createFunction(
+      'format_user_email',
```

Comment:
Can you remind me why we need this trigger again?

Reply:
this is for when prod ops (or anyone with db access) adds a user directly into the database (typically from a nonwhitelisted domain/.gov.sg account), to ensure that the email address will be normalised (i.e. all lower case)

Reply:
I thought we already have that trigger from another migration? Why are we adding it again here?

Reply:
oh so that trigger (added by the previous team) does two things:
1. ensure the user's email is lower case
2. making the edit easier by allowing the DBA to just leave `api_key` blank and the trigger would automatically convert it to `NULL`

I am adding it here to only do 1 and to NOT to do 2. That trigger will be dropped in the `part 2` PR. Other options are:
- remove the trigger entirely (i.e. don't do 1)
- replicate the functionality of that trigger (so that the DBA can leave `api_key_hash` blank and the trigger automatically convert it to null.

Haha I chose to keep 1 but drop 2 because I think: 1 is quite important to have, in case the DBA makes an error, could really cause some downstream errors; 2 is an annoying hidden coupling that's not that useful anyway.

Hope this clarifies!

<img width="619" alt="image" src="https://user-images.githubusercontent.com/67887489/196325821-efbb9604-8d4f-41c2-9aed-616338691989.png">

Reply:
Hmm but I don't see any part of this code dropping the trigger created by the previous team though 🤔

Reply:
Oh I dropped this trigger in the [next PR](https://github.com/opengovsg/postmangovsg/pull/1842/files) instead

In the unlikely even there are manual db inserts between these two PRs, the trigger should still run hahah

Reply:
Why can't we just drop the trigger from the previous team in this PR?

Reply:
Hmm yeah upon reflection ya we can drop it here. Please wait let me change the PR!